presto-product-tests/src/main/java/com/facebook/presto/tests/SqlCancelTests.java

```diff
@@ -0,0 +1,212 @@
+import io.airlift.http.client.Response;
+import io.airlift.log.Logger;
+import org.testng.annotations.Test;
+
+import java.sql.SQLException;
+import java.util.concurrent.Callable;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.Future;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+
+import static com.facebook.presto.tests.TestGroups.CANCEL_QUERY;
+import static com.teradata.tempto.assertions.QueryAssert.Row.row;
+import static com.teradata.tempto.assertions.QueryAssert.assertThat;
+import static com.teradata.tempto.query.QueryExecutor.query;
+import static java.lang.String.format;
+
+public class SqlCancelTests
+        extends ProductTest
+{
+    private static String tableName = "table_name";
+
+    @Inject
+    @Named("databases.presto.server_address")
+    private String serverAddress;
+
+    @BeforeTestWithContext
+    @AfterTestWithContext
+    public void dropTestTables()
+    {
+        try {
+            query(format("DROP TABLE IF EXISTS %s", tableName));
+        }
+        catch (Exception e) {
+            Logger.get(getClass()).warn(e, "failed to drop table");
+        }
+    }
+
+    private class SqlExecuteCallable
+            implements Callable<Boolean>
+    {
+        private String sql;
+
+        public SqlExecuteCallable(String sql)
+        {
+            this.sql = sql;
+        }
+
+        public Boolean call()
+        {
+            try {
+                query(sql);
+                return false;
+            }
+            catch (QueryExecutionException e) {
+                return true;
+            }
+        }
+    }
+
+    private class SqlCancelCallable
+            implements Callable<Response>
+    {
+        private String sqlStatement;
+        private long waitTime;
+        private long step;
+
+        public SqlCancelCallable(String sqlStatement, long waitTime, long step)
+        {
+            this.sqlStatement = sqlStatement;
+            this.waitTime = waitTime;
+            this.step = step;
+        }
+
+        public Response call()
+        {
+            long time = 0;
+            while (time < waitTime) {
+                String queryState = (String) query(format("SELECT state from system.runtime.queries" +
+                        " WHERE query = '%s' ORDER BY created DESC LIMIT 1", sqlStatement)).row(0).get(0);
+                if (!queryState.equalsIgnoreCase("running")) {
+                    try {
+                        Thread.sleep(step);
+                    }
+                    catch (InterruptedException e) {
+                        e.printStackTrace();
+                        return null;
+                    }
+                    time = time + step;
+                }
+                else {
+                    String queryId = (String) query(format("SELECT query_id from system.runtime.queries" +
```

Comment:
Instead select the queryId as part of the initial query at the beginning of the loop   (`SELECT state, query_id from system.runtime.queries ...`)